pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/rawingress_reconciler.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2021 kubeflow.org.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ingress
+
+import (
+	"context"
+
+	v1beta1api "github.com/kubeflow/kfserving/pkg/apis/serving/v1beta1"
+	"github.com/kubeflow/kfserving/pkg/constants"
+	"github.com/kubeflow/kfserving/pkg/utils"
+	corev1 "k8s.io/api/core/v1"
+	netv1 "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/apis"
+	knapis "knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+//RawIngressReconciler is the struct of Raw K8S Object
+type RawIngressReconciler struct {
+	client        client.Client
+	scheme        *runtime.Scheme
+	Ingress       *netv1.Ingress
+	URL           *knapis.URL
+	IngressConfig *v1beta1api.IngressConfig
+}
+
+func NewRawIngressReconciler(client client.Client,
+	scheme *runtime.Scheme,
+	ingressConfig *v1beta1api.IngressConfig,
+	isvc *v1beta1api.InferenceService) (*RawIngressReconciler, error) {
+	ingress, err := createRawIngress(client, scheme, isvc, ingressConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &RawIngressReconciler{
+		client:        client,
+		scheme:        scheme,
+		Ingress:       ingress,
+		URL:           createRawURL(client, isvc, ingressConfig),
+		IngressConfig: ingressConfig,
+	}, nil
+}
+func createRawURL(client client.Client,
+	isvc *v1beta1api.InferenceService,
+	ingressConfig *v1beta1api.IngressConfig) *knapis.URL {
+	url := &knapis.URL{}
+	url.Scheme = "http"
+	url.Host = isvc.Name + "-" + isvc.Namespace + "." + ingressConfig.IngressDomain
+	return url
+}
+func generateRule(podSpec *corev1.PodSpec,
+	ingressHost string,
+	componentName string) netv1.IngressRule {
+	pathType := netv1.PathType(netv1.PathTypeImplementationSpecific)
+	rule := netv1.IngressRule{
+		Host: ingressHost,
+		IngressRuleValue: netv1.IngressRuleValue{
+			HTTP: &netv1.HTTPIngressRuleValue{
+				Paths: []netv1.HTTPIngressPath{
+					{
+						PathType: &pathType,
```

Comment:
if there is explainer, we need to support both `:predict` and `:explain` which requires path based routing to predictor and explainer.

Reply:
explainer and predictor have different host, if there are explainer and predictor, the top level will point to explainer.

Reply:
there could be two verbs for explainer case and it is totally depending on the request path user specifies, when user calls `:predict` KFServing should send the request directly to predictor, when calling `:explain` path KFServing should send to the explainer.